### PR TITLE
feat: Introduce propagation with all parametrisations 

### PR DESCRIPTION
**Pull request by FabianKlimpel**
This PR is the re-opened MR https://gitlab.cern.ch/acts/acts-core/-/merge_requests/653.
It extends the concept of parameter propagation by allowing Bound, Curvilinear and FreeParameters interchangeably. 
This PR is a part of #165 

Edit:
As an introduction to this PR: Most of the changes performed here are just for compatibility, testing and API. The main changes are inside the CovarianceEngine.cpp. Since we always had covariance transport from Curvilinear (C) or Bound (B) parametrisations to C/B, there was not much to distinguish. Adding FreeParameters (F) as additional possibility we get more possible combinations, i.e. the cases C/B->F, F->C/B and F->F were added. In terms of a jacobian J from point X to point Y, we now have the possibility to have either a 8x6 (C/B->F), 6x8 (F->C/B) or 8x8 (F->F) matrix. For the covariance matrix it's a similar case. The C/B representation was given in 6x6, now 8x8 is also a possibility. Since the explicit dimensions of both aren't known prior to the propagation, a combination is stored inside the states as variant.
The individual part of the jacobian can and are stored explicitly. Since starting with FreeParameters means that there is no jacobian from local to global parameters, this becomes an optional matrix. This feature on the other hand allows to steer the workflow internally in order to receive the correct results. Another element thereby is the existence of a surface to which the propagation should be performed. Since C & F are both running without surfaces a flag was introduced.
Looking at this from a higher level, the steppers just call the corresponding functions and the CovarianceEngine performs the correct evaluations based on the given parameters internally. For the Propagator the return parametrisation was added, defaulted on C (in case of no present surface, otherwise it's B).
All other changes were just made to allow this workflow and mostly based on copy&paste of existing code with appropriate dimension adaptions or to keep the workflow as it is in the current master. In total, these changes are not visible (or at least hidden as good as possible) from the outside but provide the correct result based on the given input.

### CODE REVIEW

**Comment by msmk0** [comment]
@FabianKlimpel Can this be reviewed now?

**Comment by FabianKlimpel** [comment]
> @FabianKlimpel Can this be reviewed now?

The intended functionality is available. I might extend the unit tests a little bit further in the future to improve the coverage but that's rather optional and consequently not really WIP anymore. So yes, it can be reviewed.

File: Core/include/Acts/Fitter/KalmanFitter.hpp [hunk 1]
```diff
@@ -463,12 +463,12 @@ class KalmanFitter {
...
           state.stepping.stepSize = ConstrainedStep(state.options.maxStepSize);
           state.stepping.pathAccumulated = 0.;
+          state.stepping.jacToGlobal = BoundToFreeMatrix::Zero();
           // Reinitialize the stepping jacobian
           st.referenceSurface().initJacobianToGlobal(
-              state.options.geoContext, state.stepping.jacToGlobal,
+              state.options.geoContext, *state.stepping.jacToGlobal,
               state.stepping.pos, state.stepping.dir,
               st.filteredParameters(state.options.geoContext).parameters());
-          state.stepping.jacobian = BoundMatrix::Identity();
```
Comment (XiaocongAi) [inline Core/include/Acts/Fitter/KalmanFitter.hpp:471]: Why is `jacobian` reinitialization removed here? I expect that the reinitialization is done for each material surface, so here as well.

Comment (XiaocongAi) [reply Core/include/Acts/Fitter/KalmanFitter.hpp:471]: Ah, I see. Now, it's always `jacobian = jacFull;`.

Comment (FabianKlimpel) [reply Core/include/Acts/Fitter/KalmanFitter.hpp:471]: It's a modification of the workflow to avoid the issue that the jacobian represents the covariance transport from A to B while the covariance matrix is already transported to B. Instead the jacobian is always assigned to the last calculated one.

**Comment by FabianKlimpel** [comment]
> Thanks @FabianKlimpel for preparing this extensive PR. I finally started reviewing this; my apologies for the delay. So far I only have some general comments. I will provide more details later.
> 
> * The PR as-is is quite large; for easier reviewing it should be split up into a few smaller PRs. The event data and surface changes are good candidates to be split up into separate PRs that can be merged in first. Another option would be to first add the internal changes to the steppers w/o adding support for free parameters and then add the support for the free parameters in a second PR.

This PR is already a smaller version of itself. The event data and surface changes could be separated indeed but since these are O(10) lines there's no real benefit. Every other split (as far as I see) would lead to pointless PRs. 

> * By adding most of the functionality to the Steppers their interface becomes quite large. They are also more tightly coupled to the track parameter classes. I would prefer if this could be simplified, i.e. the stepper only provides the track parameters and the Jacobians. The covariance transport would then be handled by a new layer between the steppers and the propagator that is independent of the stepper implementation (similar to what is already done with the `CovarianceEngine`, but without hiding it inside each stepper).

A change of the general structure is possible but I will not do this within this PR, it should be kept separately.

> * Is the `std::variant<...>` implementation of the various transport really needed? If you call the propagator e.g. with free input parameters and propagate onto a surface, the type of both input and output matrices are known and will not change throughout the propagation. You should be able to select the right implementation already then and would not need to keep all possibilities as for the case of the `std::variant<...>`.

It is needed. Although you propagate to a surface you don't know what happens during the propagation. So the only thing you know is the total jacobian from start to end.

File: Core/include/Acts/Propagator/AtlasStepper.hpp [hunk 2]
```diff
@@ -495,8 +525,7 @@ class AtlasStepper {
     // prepare the jacobian if we have a covariance
     if (pars.covariance()) {
       // copy the covariance matrix
-      state.covariance =
-          new ActsSymMatrixD<eBoundParametersSize>(*pars.covariance());
+      state.covariance = new Covariance(*pars.covariance());
```
Comment (paulgessinger) [inline Core/include/Acts/Propagator/AtlasStepper.hpp:528]: `new`??

**Comment by paulgessinger** [comment]
Risking that I sound reactionary, I concur that this PR makes a ton of changes that make it hard to follow individual decisions. I also would be careful with `std::variant` in this context, the code is now littered with `std::get`, which performs checks against the type and potentially throws. 

As a general comment, I feel like the propagation component is becoming more and more of a beast, the implementation becomes increasingly convoluted and hard to follow, readability takes a turn for the worse. I have doubts that anyone other than @FabianKlimpel could dive in and make any real changes at this point, and while his work is excellent, I'm not sure that's a good thing overall.

**Comment by FabianKlimpel** [comment]
As an introduction to this PR: Most of the changes performed here are just for compatibility, testing and API. The main changes are inside the `CovarianceEngine.cpp`. Since we always had covariance transport from Curvilinear (C) or Bound (B) parametrisations to C/B, there was not much to distinguish. Adding FreeParameters (F) as additional possibility we get more possible combinations, i.e. the cases C/B->F, F->C/B and F->F were added. In terms of a jacobian J from point X to point Y, we now have the possibility to have either a 8x6 (C/B->F), 6x8 (F->C/B) or 8x8 (F->F) matrix. For the covariance matrix it's a similar case. The C/B representation was given in 6x6, now 8x8 is also a possibility. Since the explicit dimensions of both aren't known prior to the propagation, a combination is stored inside the states as variant.
The individual part of the jacobian can and are stored explicitly. Since starting with FreeParameters means that there is no jacobian from local to global parameters, this becomes an optional matrix. This feature on the other hand allows to steer the workflow internally in order to receive the correct results. Another element thereby is the existence of a surface to which the propagation should be performed. Since C & F are both running without surfaces a flag was introduced.
Looking at this from a higher level, the steppers just call the corresponding functions and the CovarianceEngine performs the correct evaluations based on the given parameters internally. For the Propagator the return parametrisation was added, defaulted on C (in case of no present surface, otherwise it's B).
All other changes were just made to allow this workflow and mostly based on copy&paste of existing code with appropriate dimension adaptions or to keep the workflow as it is in the current master. In total, these changes are not visible (or at least hidden as good as possible) from the outside but provide the correct result based on the given input.

File: Core/include/Acts/Propagator/detail/CovarianceEngine.hpp [hunk 3]
```diff
@@ -75,14 +96,41 @@ std::tuple<BoundParameters, BoundMatrix, double> boundState(
...
 ///   - and the path length (from start - for ordering)
-std::tuple<CurvilinearParameters, BoundMatrix, double> curvilinearState(
-    BoundSymMatrix& covarianceMatrix, BoundMatrix& jacobian,
+std::tuple<CurvilinearParameters, Jacobian, double> curvilinearState(
+    Covariance& covarianceMatrix, Jacobian& jacobian,
     FreeMatrix& transportJacobian, FreeVector& derivatives,
-    BoundToFreeMatrix& jacobianLocalToGlobal, const FreeVector& parameters,
+    std::optional<BoundToFreeMatrix>& jacobianLocalToGlobal,
+    ActsMatrixD<8, 7>& jacobianDirToAngle,
+    ActsMatrixD<7, 8>& jacobianAngleToDir, const FreeVector& parameters,
```
Comment (XiaocongAi) [inline Core/include/Acts/Propagator/detail/CovarianceEngine.hpp:104]: Could you add comments for the `jacobianDirAngle` and `jacobianAngleToDir`?

File: Core/include/Acts/Propagator/detail/CovarianceEngine.hpp [hunk 4]
```diff
@@ -24,6 +24,25 @@ namespace Acts {
...
+using Jacobian =
+    std::variant<BoundMatrix, FreeToBoundMatrix, BoundToFreeMatrix, FreeMatrix>;
+
+/// @brief Constructs a jacobian to transform from (x,y,z,t,phi,theta,q/p) to
+/// (x,y,z,t,Tx,Ty,Tz,q/p)
+///
+/// @param [in] dir Direction vector
+///
+/// @return The jacobian
+ActsMatrixD<8, 7> jacobianDirectionsToAngles(const Vector3D& dir);
```
Comment (XiaocongAi) [inline Core/include/Acts/Propagator/detail/CovarianceEngine.hpp:37]: Maybe I misunderstand something, but I guess the name of this function should be `jacobianAnglesToDirections`...

File: Core/include/Acts/Propagator/detail/CovarianceEngine.hpp [hunk 5]
```diff
@@ -24,6 +24,25 @@ namespace Acts {
...
+
+/// @brief Constructs a jacobian to transform from (x,y,z,t,Tx,Ty,Tz,q/p) to
+/// (x,y,z,t,phi,theta,q/p)
+///
+/// @param [in] dir Direction vector
+///
+/// @return The jacobian
+ActsMatrixD<7, 8> jacobianAnglesToDirections(const Vector3D& dir);
```
Comment (XiaocongAi) [inline Core/include/Acts/Propagator/detail/CovarianceEngine.hpp:45]: And the name of this function should be `jacobianDirectionsToAngles`?

File: Core/include/Acts/Propagator/detail/CovarianceEngine.hpp (hunk 4 above)
Comment (FabianKlimpel) [reply Core/include/Acts/Propagator/detail/CovarianceEngine.hpp:37]: The naming is a bit misleading, you're right. Since it is a transformation that is multiplied from the right side to the jacobian it actually transforms the dependency as the naming suggests. I will think about renaming them.

File: Core/include/Acts/Propagator/detail/CovarianceEngine.hpp (hunk 3 above)
Comment (FabianKlimpel) [reply Core/include/Acts/Propagator/detail/CovarianceEngine.hpp]: Fixed

File: Core/include/Acts/Propagator/detail/CovarianceEngine.hpp (hunk 5 above)
Comment (FabianKlimpel) [reply Core/include/Acts/Propagator/detail/CovarianceEngine.hpp]: Fixed

**Comment by paulgessinger** [comment]
Github is struggling with this PR since it has over 300 commits. Maybe you could squash some of them?

**Comment by FabianKlimpel** [comment]
> Github is struggling with this PR since it has over 300 commits. Maybe you could squash some of them?

Thanks for pointing that out. It seems like I've doubled the commits.
